### PR TITLE
Add client side block filter (BIP158)

### PR DESCRIPTION
**Pull request by tamasblummer**
Code moved here from murmel. Has been running there fore a while.

@stevenroose dependency on siphasher is temporary. Will use bitcoin_hashes if https://github.com/rust-bitcoin/bitcoin_hashes/pull/46 is merged.

### CODE REVIEW

**Comment by tamasblummer** [comment]
@apoelstra this fails to build on 1.22 because of missing u128 support. 1.26 would be sufficient. How long do you plan to stick at 1.22 ?

File: src/util/blockfilter_test.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter_test.json]: yes, but pretty big though. It might make sense to have an extra directory for test files.

File: src/util/blockfilter.rs [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+//!
+
+use blockdata::{
+        transaction::OutPoint,
+        block::Block,
+        script::Script
+    };
+use consensus::{Decodable, Encodable, encode::VarInt};
+use util::hash::BitcoinHash;
+use bitcoin_hashes::sha256d::Hash as Sha256dHash;
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter.rs]: Yes, this was a quick port into rust-bitcoin. This applies to most of your comments.

File: src/util/blockfilter.rs [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+    pub block: Sha256dHash,
+    /// filte type (see SCRIPT_FILTER)
+    pub filter_type: u8,
+    /// Golomb encoded filter
+    pub content: Vec<u8>
+}
+
+impl BlockFilter {
+    /// Compute a SCRIPT_FILTER that contains spent and output scripts
+    pub fn compute_script_filter(block: &Block, scripts: Box<ScriptAccessor>) -> Result<BlockFilter, Box<Error>> {
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter.rs]: I do not like non-trivial closures, there are ugly examples for that pattern eg. in tokio. The Box<dyn Trait> is less cumbersome and the call overhead does not matter as it is called once to resolve all coins referred, which is likely optimal for a db storing utxos.

[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter.rs]: There is an error type there already, that could be extended. I do not think it adds value to re-wrap e.g. io errors, but you are right that is the pattern in the rest of the rust-bitcoin code.

[the remark this answers is not in the document]
Comment (apoelstra) [reply src/util/blockfilter.rs]: Using boxes means fat pointers, allocations, indirections, `std` dependency, and it prevents users from matching on the actual error type.

[the remark this answers is not in the document]
Comment (apoelstra) [reply src/util/blockfilter.rs]: Could you at least parameterize over `ScriptAccessor` instead of using a `Box`? Every allocation is an `std` dependency and contributes to memory fragmentation and requires users to type `Box::new`.

[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter.rs]: I can make this a closure if you prefer. I understand implications of using Box, but is it really an issue here given that the retrieved objects are Scripts from TxOuts that were defined by you as Box<[u8]> ?

**Comment by tamasblummer** [comment]
@apoelstra @stevenroose addressed all your comments except replacement of Box<dyn Trait> with a closure. Squash or do you insist on that?

File: src/util/blockfilter_test.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter_test.json]: I think that the API is rather precise, documented in module comments.

File: src/util/blockfilter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter.rs]: I think the API is fine. Added use comments in module header.

**Comment by tamasblummer** [comment]
BLOCKED on rust-bitcoin/bitcoin_hashes#46 wich would replace siphasher used here which is using u128 which is not supported below rustc 1.26

File: src/util/blockfilter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/util/blockfilter.rs]: If you just do it like this:
```rust
     pub fn compute_script_filter<A: ScriptAccessor>(block: &Block, scripts: A) -> Result<BlockFilter, Box<Error>>
```
Then one can implement `impl ScriptAccessor for Box<Whatever>`, right? Or f.e. `impl<'a> ScriptAccessor for &'a Whatever`. Then the decision to heap-allocate is on the user.

**Comment by stevenroose** [comment]
https://github.com/rust-bitcoin/bitcoin_hashes/pull/46 is merged but not yet released.

Your `u128` method can be replaced with this, just like btcsuite does here: https://github.com/btcsuite/btcutil/blob/master/gcs/gcs.go#L54

You could perhaps even have two implementations depending on whether `u128` is supported, not sure it's possible to build over that condition.

About the API, at first, I was confused with the Writer and Reader thing. I think the Writer is not really needed, but I agree the Reader can help prevent some unnecessary allocations.

Could you perhaps also add the match methods to the `BlockFilter` type and just pass them on to a Reader-wrapper?

File: src/util/bip158.rs [hunk 4]
```diff
@@ -190,13 +199,13 @@ impl <'a> BlockFilterWriter<'a> {
 
 
 /// Reads and interpret a block filter
-pub struct BlockFilterReader {
```
Comment (stevenroose) [inline src/util/bip158.rs]: Oh, why did you make these ones private? I thouht the Reader was so that a user could use the match method without having to re-allocate the bytes. Like this they still have to. Leave these public and the user has the choice to just allocate a `BlockFilter` or to wrap their existing bytes with the Reader.

Comment (tamasblummer) [reply src/util/bip158.rs]: ok

**Comment by tamasblummer** [comment]
parametrized BlockFilter and moved match methods into it, that is the only API now.
I prefer the u128 replacement I wrote, the compiler should do the optimization if there is any. I looked up but could not find conditional compilation on rustc version without added dev dependency. Is it worth?

File: src/util/bip158.rs (hunk 4 above)
Comment (tamasblummer) [reply src/util/bip158.rs]: wait. there is no savings there. nothing can be reused

**Comment by stevenroose** [comment]
Ah ok cool you already wrote a non-u128 version, I didn't see that in your update. No that should do! Did you read my comment about the match api on both blockfilter and blockfilterreader?

Btw, is there any reason why `BlockFilter(Reader|Writer)` is split out from `GCSFilter(Reader|Writer)`?

**Comment by tamasblummer** [comment]
GCSFilterRW are more generic, BlockFilterRWs  are assuming that sip keys are derived from block hash. Squash?

**Comment by stevenroose** [comment]
I'd first make the Reader methods public again :) Then squash.

File: src/util/bip158.rs (hunk 4 above)
Comment (tamasblummer) [reply src/util/bip158.rs]: made public again

**Comment by tamasblummer** [comment]
@stevenroose squashed. Using patch in Cargo.toml to access master of bitcoin_hashes. @apoelstra please do a release of bitcoin_hashes, so siphash24 can be used here.

File: src/util/blockfilter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/blockfilter.rs]: parametrized

**Comment by apoelstra** [comment]
We need to backport siphasher to 0.3.x of bitcoin_hashes that includes siphasher. We will need to do a backport release anyway. I'll try to do this today.

**Comment by apoelstra** [comment]
Actually we could just add a commit to this PR which updates to bitcoin_hashes 0.4.

**Comment by tamasblummer** [comment]
@apoelstra This already updates to bitcoin_hashes 0.4

**Comment by tamasblummer** [comment]
@stevenroose I think all changes you requested were addressed.

**Comment by stevenroose** [comment]
@tamasblummer I know, I know. It's just quite a chunk to process. I'm definitely planning to give it a review!

File: src/lib.rs [hunk 5]
```diff
@@ -86,3 +86,4 @@ pub use util::key::PrivateKey;
 pub use util::key::PublicKey;
 pub use util::decimal::Decimal;
 pub use util::decimal::UDecimal;
+pub use util::bip158;
```
Comment (stevenroose) [inline src/lib.rs]: We don't do this for any other modules in `util`, so I don't think you should. Users can do `use bitcoin::util::bip158` and then use `bip158::BlockFilter` etc.

File: src/util/bip158.rs [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+            let mut writer = BlockFilterWriter::new(&mut out, block);
+            writer.add_output_scripts();
+            writer.add_input_scripts(script_for_coin)?;
+            writer.finish()?;
+        }
+        Ok(BlockFilter{block: block.bitcoin_hash(), filter_type: SCRIPT_FILTER, content: out.into_inner()})
+    }
+
+    /// match any query pattern
+    pub fn match_any (block_hash: &sha256d::Hash, reader: &mut io::Read, query: &mut Iterator<Item=&[u8]>) -> Result<bool, io::Error> {
```
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/util/bip158.rs]: If you change this, don't forget to also change the example in the module documentation.

File: src/util/bip158.rs [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+use std::{cmp,fmt,io};
+use std::io::Cursor;
+use std::error::Error;
+use std::fmt::{Debug, Display, Formatter};
+use std::collections::HashSet;
+
+/// BIP158 base filter type 0: input and output scripts
+pub const SCRIPT_FILTER:u8 = 0;
+
+const P: u8 = 19;
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/bip158.rs]: c'mon what else. the whole file is an implementation of BIP-158

File: src/util/bip158.rs [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+use blockdata::block::Block;
+use blockdata::script::Script;
+
+use consensus::{Decodable, Encodable};
+use consensus::encode::VarInt;
+use util::hash::BitcoinHash;
+use bitcoin_hashes::{sha256d, Hash, siphash24};
+use byteorder::{ByteOrder, LittleEndian};
+
+use std::{cmp,fmt,io};
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/bip158.rs]: run format

File: src/util/bip158.rs [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+    pub block: sha256d::Hash,
+    /// filte type (see SCRIPT_FILTER)
+    pub filter_type: u8,
+    /// Golomb encoded filter
+    pub content: Vec<u8>
+}
+
+impl BlockFilter {
+    /// Compute a SCRIPT_FILTER that contains spent and output scripts
+    pub fn compute_script_filter<M>(block: &Block, script_for_coin: M) -> Result<BlockFilter, BlockFilterError>
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/bip158.rs]: renamed to new_script_filter

File: src/util/bip158.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/bip158.rs]: Thanks for these observation. I cleaned this up.

File: src/util/bip158.rs [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+    use rand;
+    use rand::Rng;
+    use std::collections::{HashSet, HashMap};
+    use std::io::Cursor;
+    use bitcoin_hashes::hex::FromHex;
+    use super::*;
+
+    extern crate hex;
+
+    fn decode<T: ? Sized>(data: Vec<u8>) -> Result<T, io::Error>
```
[the remark this answers is not in the document]
Comment (tamasblummer) [reply src/util/bip158.rs]: no, this is the reverse operation

**Comment by tamasblummer** [comment]
@stevenroose addressed all your comments, re-based, squashed

File: src/util/bip158.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/util/bip158.rs]: Hmm yeah that's a valid point, I guess :)

File: src/util/bip158.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (stevenroose) [reply src/util/bip158.rs]: Oh sorry, you're right. Even though there is `bitcoin::consensus::encode::deserialize`, it's not the biggest deal to have this in the test module.

**Comment by stevenroose** [comment]
Heh, is appears that you were doing the special block thing for a reason, as it fails to compile for rust 1.22 (see Travis). It seems that 1.22 doesn't realize the mutable borrow ends with `finish` consuming the `writer` variable.

```
error[E0505]: cannot move out of `out` because it is borrowed
   --> src/util/bip158.rs:133:75
    |
127 |         let mut writer = BlockFilterWriter::new(&mut out, block);
    |                                                      --- borrow of `out` occurs here
...
133 |         Ok(BlockFilter { block_hash, filter_type: SCRIPT_FILTER, content: out.into_inner(), filter_reader })
    |                                                                           ^^^ move out of `out` occurs here
error: aborting due to previous error
```

File: src/util/bip158.rs [hunk 11]
```diff
@@ -355,7 +355,9 @@ impl<'a> GCSFilterWriter<'a> {
 
     /// Add some data to the filter
     pub fn add_element(&mut self, element: &[u8]) {
-        self.elements.insert(element.to_vec());
+        if element.len () > 0 {
```
Comment (stevenroose) [inline src/util/bip158.rs]: Space between `len` and `()` (cargo fmt? :D)
Using `if !elements.is_empty() {` is more clean, though.

File: src/util/bip158.rs [hunk 12]
```diff
@@ -526,61 +528,48 @@ mod test {
...
-            let mut txmap = HashMap::new();
-            let ref test_case = blocks[t];
-            let block_hash = sha256d::Hash::from_hex(test_case[1]).unwrap();
-            let previous_header_hash = sha256d::Hash::from_hex(test_case[3]).unwrap();
-            let header_hash = sha256d::Hash::from_hex(test_case[5]).unwrap();
-            let block: Block = deserialize(hex::decode(test_case[2]).unwrap().as_slice()).unwrap();
+
+        // test vectors from: https://github.com/jimpo/bitcoin/blob/c7efb652f3543b001b4dd22186a354605b14f47e/src/test/data/blockfilters.json
+        //["Block Height,Block Hash,Block,[Prev Output Scripts for Block],Previous Basic Header,Basic Filter,Basic Header,Notes"]
+        let data = "[\
```
Comment (stevenroose) [inline src/util/bip158.rs]: You can make the string with `r#" json here "#` and you won't have to escape all the double quotes.
https://rahul-thakoor.github.io/rust-raw-string-literals/

File: src/util/bip158.rs (hunk 11 above)
Comment (tamasblummer) [reply src/util/bip158.rs]: modified

File: src/util/bip158.rs (hunk 12 above)
Comment (tamasblummer) [reply src/util/bip158.rs]: modified

**Comment by tamasblummer** [comment]
@stevenroose sorry, I found some edge cases in added tests. please re-approve.
